storage/aptosdb/src/db_debugger/truncate/mod.rs

```diff
@@ -0,0 +1,567 @@
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        for version in start_version..end_version {
+            batch.delete::<TransactionInfoSchema>(&version)?;
+            batch.delete::<TransactionSchema>(&version)?;
+            batch.delete::<VersionDataSchema>(&version)?;
+            batch.delete::<WriteSetSchema>(&version)?;
+        }
+
+        Ok(())
+    }
+
+    fn delete_state_value_and_index(
+        ledger_db: &DB,
+        start_version: Version,
+        end_version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()> {
+        let mut iter = ledger_db.iter::<StaleStateValueIndexSchema>(ReadOptions::default())?;
+        iter.seek(&start_version)?;
+
+        for item in iter {
+            let (index, _) = item?;
+            assert_lt!(index.stale_since_version, end_version);
+            batch.delete::<StaleStateValueIndexSchema>(&index)?;
+            batch.delete::<StateValueSchema>(&(index.state_key, index.stale_since_version))?;
+        }
+
+        Ok(())
+    }
+
+    fn truncate_state_merkle_db(state_merkle_db: &DB, target_version: Version) -> Result<()> {
+        let status = StatusLine::new(Progress::new(target_version));
+        loop {
+            let batch = SchemaBatch::new();
+            let current_version = Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+            status.set_current_version(current_version);
+            assert_ge!(current_version, target_version);
+            if current_version == target_version {
+                break;
+            }
+
+            let mut iter =
+                state_merkle_db.iter::<JellyfishMerkleNodeSchema>(ReadOptions::default())?;
+            iter.seek(&NodeKey::new_empty_path(current_version))?;
+            for item in iter {
+                let (key, _) = item?;
+                batch.delete::<JellyfishMerkleNodeSchema>(&key)?;
+            }
+
+            Self::delete_stale_node_index_at_version::<StaleNodeIndexSchema>(
+                state_merkle_db,
+                current_version,
+                &batch,
+            )?;
+            Self::delete_stale_node_index_at_version::<StaleNodeIndexCrossEpochSchema>(
+                state_merkle_db,
+                current_version,
+                &batch,
+            )?;
+
+            state_merkle_db.write_schemas(batch)?;
+        }
+
+        Ok(())
+    }
+
+    fn delete_stale_node_index_at_version<S>(
+        state_merkle_db: &DB,
+        version: Version,
+        batch: &SchemaBatch,
+    ) -> Result<()>
+    where
+        S: Schema<Key = StaleNodeIndex>,
+        Version: SeekKeyCodec<S>,
+    {
+        let mut iter = state_merkle_db.iter::<S>(ReadOptions::default())?;
+        iter.seek(&version)?;
+        for item in iter {
+            let (index, _) = item?;
+            assert_eq!(index.stale_since_version, version);
+            batch.delete::<S>(&index)?;
```

Comment:
is this delete here supposed to make the nodes not stale anymore after the truncation?

Reply:
The meaning of the stale index is: this node is becoming stale since the `index.stale_since_version` (because we write a new node at this version). If this version is larger than target_version, it should be deleted.